sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala

```diff
@@ -68,10 +71,56 @@ object InferWindowGroupLimit extends Rule[LogicalPlan] with PredicateHelper {
     case _ => false
   }
 
+  /**
+   * All window expressions should not have SizeBasedWindowFunction, all lower/upper of
+   * specifiedWindowFrame is UnboundedPreceding/CurrentRow, and window orderSpec is not foldable,
+   * so that we can safely do the early stop.
+   */
+  private def limitSupport(limit: Int, window: Window): Boolean =
+    limit <= conf.windowGroupLimitThreshold && window.child.maxRows.forall(_ > limit) &&
+      !window.child.isInstanceOf[WindowGroupLimit] &&
+      window.orderSpec.exists(!_.foldable) &&
+      !LimitPushDownThroughWindow.supportsPushdownThroughWindow(window.windowExpressions) &&
+      window.windowExpressions.forall {
+        case Alias(WindowExpression(windowFunction, WindowSpecDefinition(_, _,
+        SpecifiedWindowFrame(_, UnboundedPreceding, CurrentRow))), _)
+          if !windowFunction.isInstanceOf[SizeBasedWindowFunction] => true
+        case _ => false
+      }
+
+  private def isRowFrame(windowExpression: NamedExpression): Boolean = windowExpression match {
+    case Alias(WindowExpression(_, WindowSpecDefinition(_, _,
+    SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow))), _) => true
+    case _ => false
+  }
+
+  private def rankLikeFunction(windowExpressions: Seq[NamedExpression]): Expression =
+    // If windowExpressions all are RowFrame, choose SimpleLimitIterator,
+    // else RankLimitIterator to obtain enough rows for ensure data accuracy.
+    if (windowExpressions.forall(isRowFrame)) {
+      new RowNumber
+    } else {
+      new Rank
```

Comment:
requires all the frame is `RangeFrame` ?

Reply:
Take an example, aggregate function `sum(a) with row frame` and another `sum(a) with range frame`.
Now we select the `Rank`.
Is the output of `sum(a) with row frame` correct?

Reply:
There is a problem with this example. Another example, `df.withColumn("rn", row_number().over(window)).withColumn("sum_value", sum("value").over(window)).limit(1) `. row_number is row frame, sum is range frame, We need to select `RankLimitIterator`. If select `SimpleLimitIterator`, sum value data will be wrong when the same row exists.

Reply:
`df.withColumn("sum_value1", sum("value").over(window.rowsBetween(Window.unboundedPreceding, Window.currentRow))).withColumn("sum_value2", sum("value").over(window)).limit(1)` ?

Reply:
They are two window.

Reply:
No. split Window only with partition spec, order spec and window function type. The window frame is not participant.

`df.withColumn("sum_value1", sum("value").over(window.rowsBetween(Window.unboundedPreceding, Window.currentRow))).withColumn("sum_value2", sum("value").over(window.rangeBetween(Window.unboundedPreceding, Window.currentRow))).limit(1)`

Reply:
Yes. `sum_value2` will be correct is select `RankLimitIterator`.
But what I want know is the result of `sum_value1`.

Reply:
`sum_value1` not equal `sum_value2` if exist same rows.

Reply:
`WindowGroupLimit` only limits the partition and does not perform value calculation. The value calculation is still done by `Window`, so we only need to ensure the security of `WindowGroupLimit`.